workflow/sync/syncitems.go

```diff
@@ -0,0 +1,52 @@
+package sync
+
+import (
+	"errors"
+	"reflect"
+
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+type syncItem struct {
+	semaphore *v1alpha1.SemaphoreRef
+	mutex     *v1alpha1.Mutex
+}
+
+func allSyncItems(sync *v1alpha1.Synchronization) ([]*syncItem, error) {
+	var syncItems []*syncItem
+	if sync.Semaphore != nil {
+		syncItems = append(syncItems, &syncItem{semaphore: sync.Semaphore})
+	}
+	if sync.Mutex != nil {
+		syncItems = append(syncItems, &syncItem{mutex: sync.Mutex})
+	}
+	for _, semaphore := range sync.Semaphores {
+		syncItems = append(syncItems, &syncItem{semaphore: semaphore})
+	}
+	for _, mtx := range sync.Mutexes {
+		syncItems = append(syncItems, &syncItem{mutex: mtx})
+	}
+	return syncItems, checkDuplicates(syncItems)
+}
+
+func checkDuplicates(items []*syncItem) error {
+	for i, item := range items {
+		for j := i + 1; j < len(items); j++ {
+			if reflect.DeepEqual(*item, *items[j]) {
+				return errors.New("Duplicate synchronization item found")
+			}
+		}
+	}
+	return nil
+}
+
+func (i *syncItem) getType() v1alpha1.SynchronizationType {
+	switch {
+	case i.semaphore != nil:
+		return v1alpha1.SynchronizationTypeSemaphore
+	case i.mutex != nil:
+		return v1alpha1.SynchronizationTypeMutex
+	default:
+		return v1alpha1.SynchronizationTypeUnknown
```

Comment:
I did consider it. I didn't introduce `SynchronizationTypeUnknown`, and it would mean a change to [GetStatus](https://github.com/argoproj/argo-workflows/blob/22b64f98998754216e56bbf768fc66f92ecb0f5e/pkg/apis/workflow/v1alpha1/workflow_types.go#L3878) which I'm often calling as `GetStatus(getType())` so I tried not to refactor too far.
Can we do this in another PR instead to avoid scope creep.

Reply:
Yeah okay in that case, this makes sense for a later PR